### ISSUE TITLE
Multiple documents in a stream

### ISSUE THREAD

**Issue by fabriziocucci** [FasterXML/jackson-dataformat-yaml#72]
Hi,

sorry in advance if this was asked before but I was wondering if the possibility to deserialize multiple documents in a single stream has been considered before.

From the [YAML specification](http://www.yaml.org/spec/1.2/spec.html#id2760395), the following:

``` yaml
# Ranking of 1998 home runs

---
- Mark McGwire
- Sammy Sosa
- Ken Griffey

# Team ranking

---
- Chicago Cubs
- St Louis Cardinals
```

is a valid stream containing two documents which could be (potentially) deserialized as an array of two elements, each of which would be an array of, respectively, three and two elements.

Currently (with jackson-dataformat-yaml-2.8.2), the following code:

``` java
String yaml = String.join(System.getProperty("line.separator")
                , "---"
                , "- inDocument1"
                , "---"
                , "- inDocument2");
ObjectMapper yamlObjectMapper = new ObjectMapper(new YAMLFactory());
JsonNode jsonNode = yamlObjectMapper.readValue(yaml, JsonNode.class);
System.out.println(jsonNode);
```

only prints the first document.

**Comment by cowtowncoder** [comment]
@fabriziocucci I don't think this has been asked before.

From API perspective, such use is supported, but not via simple `readValue()`, as the result is not valid as one of JSON datatypes (Object, Array, scalar).
But it may be read using `MappingIterator` via `ObjectReader`; f.ex:

```java
ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
MappingIterator<JsonNode> it = mapper.readerFor(JsonNode.class)
    .readValues(yaml);
while (it.hasNextValue()) {
  JsonNode tree = it.nextValue();
  // process
}
```

Now... I have not tried this with YAML, so it may be that this does not currently work.
If so, we'd have to figure out how to make it work; and I think that's a reasonable thing to do as well.
It just means that `YAMLParser` has to make sure its usage of SnakeYAML allows decoding of more tokens -- I don't know if and what extra work is needed. Perhaps stream of tokens exposes another document.

**Comment by fabriziocucci** [comment]
@cowtowncoder thanks for your prompt reply.

I actually didn't know about the `MappingIterator` and I agree with you on the API.

I just tried the following code:

```java
String yaml = String.join(System.getProperty("line.separator")
                , "---"
                , "- inDocument1"
                , "---"
                , "- inDocument2");
ObjectMapper yamlObjectMapper = new ObjectMapper(new YAMLFactory());
MappingIterator<JsonNode> mappingIterator = yamlObjectMapper.readerFor(JsonNode.class).readValues(yaml);
while (mappingIterator.hasNextValue()) {
	System.out.println(mappingIterator.nextValue());
}
```

and, unfortunately, it doesn't work!

**Comment by cowtowncoder** [comment]
@fabriziocucci Ok. Yes, it should be implemented; and on writer side similarly `SequenceWriter` should be usable for writing multi-document output, if possible.

I wonder how easy this would be to implement, but conceptually it seems to make sense, as each document is, I think, a single "value", usually Object.

**Comment by fabriziocucci** [comment]
@cowtowncoder I was poking around in debug just to have an idea of the work necessary to support this.

In my fantastic world of fairy tales, I was imagining the `MappingIterator` to be an interface with a specific implementation for yaml (e.g. `YamlMappingIterator`), wrapping the SnakeYaml iterator (i.e. `new Yaml().loadAll(yamlStream).iterator()`) and simply converting each SnakeYaml document  into a Jackson `JsonNode`.

But then I realized that current implementation is way more complex than that and I will need to go deep into the current implementation! :sweat_smile:

As a second point, I noticed something a bit inconsistent on the JSON part: if the documents are objects the `MappingIterator` works as expected but if the documents are arrays it doesn't.

The following works as expected:

```java
public void testThatStreamContainingTwoJsonObjectsIsProperlyParsedWithJackson() throws Exception {
	
	String json = String.join(System.getProperty("line.separator"), 
			"{ \"document1\": 1 }", 
			"{ \"document2\": 2 }");
	
	ObjectMapper jsonObjectMapper = new ObjectMapper();
	MappingIterator<JsonNode> mappingIterator = jsonObjectMapper.readerFor(JsonNode.class).readValues(json);
	while (mappingIterator.hasNextValue()) {
		System.out.println(mappingIterator.nextValue());
	}
}
```

The following doesn't:

```java
public void testThatStreamContainingTwoJsonArraysIsProperlyParsed() throws JsonProcessingException, IOException {
	
	String json = String.join(System.getProperty("line.separator"), 
			"[ 1 ]", 
			"[ 2 ]");
	
	ObjectMapper jsonObjectMapper = new ObjectMapper();
	MappingIterator<JsonNode> mappingIterator = jsonObjectMapper.readerFor(JsonNode.class).readValues(json);
	while (mappingIterator.hasNextValue()) {
		System.out.println(mappingIterator.nextValue());
	}
}
```

Is it something expected?

**Comment by cowtowncoder** [comment]
@fabriziocucci Oh, I may have mislead you there: you absolutely shouldn't have to do anything at `jackson-databind` level: that is completely format-agnostic. Rather it's question of how to make `YAMLParser` gracefully handle multi-document input, and expose token stream that also has more than one `START_OBJECT, ..... , END_OBJECT` value.

As to Object vs Array, it should work similarly. If not, please file an issue for `jackson-databind`. There are some potential problems with the fact that `MappingIterator` is designed to work not just root-level sequences, but also for elements of (JSON) Arrays. This becomes problematic if there is no way to know whether you might just have a single array of values (in which case iterator ought to expose values contains), or a sequence of arrays (in which case it should expose those arrays).
So there must be at least one case where handling has to choose one of two ambiguous cases.

**Comment by fabriziocucci** [comment]
@cowtowncoder thanks for the detailed answer on the first point, I'll try to have look!

As for the second point, I just filed an [issue](https://github.com/FasterXML/jackson-databind/issues/1377) for `jackson-databind` as you suggested.

**Comment by xiGUAwanOU** [comment]
Hi,

I've come up with an naive idea to fix this problem.

I've looked into the `MappingIterator` class, and one of the conditions to close a stream is to have one `null` returned by `getCurrentToken` method and another one `null` returned by `nextToken` method of a parser.

```text
num: 42 (*1)
---(*2)
num: -42
```

In the YAMLParser implementation, once it comes to the end of the first object ((*1) shown above), there will be an `END_OBJECT` token returned. And then, somehow the `clearCurrentToken()` method in `ParserMinimalBase` will be called (I didn't dig into the code for the reason), the `_currToken` will be set to `null`. This will make `getCurrentToken` return `null`, and this will become the first `null` in the closing stream condition.

And then the parser hits the splitter of the multiple documents ((*2) shown above), which will be treated as `DocumentEnd` by the SnakeYAML, and `YAMLParser` will return `null` on this, which actually becomes the second `null` for the closing stream condition, so that the stream is closed.

The solution is just ignore the `DocumentEnd` event from the SnakeYAML, which will make `MappingIterator` continue with the input stream. I've run the unit tests, there are 3 failed, however they have already failed before I've started editing the code, so I guess this won't be a problem?

You could review [this commit](https://github.com/xiGUAwanOU/jackson-dataformat-yaml/commit/dd5af9a3e61fa6d8b4f037aed8f090f3abea9117) to see what I've done.

**Comment by cowtowncoder** [comment]
@xiGUAwanOU Thank you, this should help. I agree that maybe swallowing end-document makes sense, but I will need to have a look at failures.

**Comment by cowtowncoder** [comment]
I think that's a reasonable fix so I merged that and accompanying test.
Will be included in 2.8.8 / 2.9.0

**Comment by xiGUAwanOU** [comment]
Thanks! Happy to see that my change gets accepted.
